### ISSUE TITLE
`uv python pin` should support writing to a `uv.toml` or `pyproject.toml`

### ISSUE THREAD

**Issue by zanieb** [astral-sh/uv#4970]
We need to design the schema for this data. Probably at `uv.python-version` or `uv.python.default-version`.

**Comment by T-256** [comment]
Then, should it also be top-level command `uv pin`? since it tries to manage project/workspace directly.

**Comment by zanieb** [comment]
That's a reasonable question. I'm not sure. I don't love it at the top-level since it's not obvious what we're pinning.

**Comment by zanieb** [comment]
I duped myself, https://github.com/astral-sh/uv/issues/4359 covers reading from the configuration file.

**Comment by charliermarsh** [comment]
(At least figure out the rough design prior to the release.)

**Comment by edmorley** [comment]
I'm slightly concerned this will lead to more fragmentation of where third party ecosystems (such as Heroku, GHA setup-python, Dependabot, ...) will need to check for Python versions when bootstrapping an environment.

If there was a field in `pyproject.toml` that was guaranteed in the upstream spec to be a single Python version (either `X.Y.Z` or `X.Y`) rather than a range, then I'd be fine with the ecosystem consolidating around that, but as-is the existing `pyproject.toml` `requires-python` field isn't suitable for bootstrapping purposes, and so the `.python-version` file is the next best thing available - and pretty widely accepted at this point.

**Comment by zanieb** [comment]
> ... as-is the existing pyproject.toml requires-python field isn't suitable for bootstrapping purposes, and so the .python-version file is the next best thing available - and pretty widely accepted at this point.

Agreed. That's why we've delayed implementing this. Unfortunately, we're getting quite a bit of confusion from users. I'm not sure how we'll address that.

**Comment by trondhindenes** [comment]
just to opine on this: The number one issue I find when helping fellow python devs un-tangling their botched python dev setups, is the presence of a ".python-version" file in some directory, since it has potential far-reaching effects. If the user happens to have "pyenv" installed, then this file in the current (or parent) dir(s) will replace the default python interpreter, leading to expected commands not being available etc. 

It is therefore problematic for uv to create and/or use this file, as it has side-efects far outside of uv itself (think developer just cding into a dir in their terminal in order to run some command). 
The fact that ".python-version" is "widely acceptable" is the exact reason uv should _not_ go anywhere near it - it should aim to minimize the risk affecting other aspects of a developer's environment. I would ask maintaines to consider at least supporting that an alternate filename like `.uv.python-version` or similar is supported as a fallback if ".python-version" is not found.
